lib/resultuploaders/s3compatible/s3compatible.rb

```diff
@@ -0,0 +1,199 @@
+# frozen_string_literal: true
+
+require 'aws-sdk-s3'
+require './lib/auxiliary/json_helper'
+
+# AutoHCK module
+module AutoHCK
+  # S3compatible class for managing interactions with S3-compatible storage services.
+  # This class is designed to work with services that are compatible with the S3 API,
+  # including AWS S3 and Alibaba Cloud OSS.
+  class S3compatible
+    CONFIG_JSON = 'lib/resultuploaders/s3compatible/s3compatible.json'
+    INDEX_FILE_NAME = 'index.html'
+
+    include Helper
+
+    # Provides read access to the URL for the index.html file.
+    #
+    # In the context of S3 compatibility, there is no concept of folders.
+    # The `url` will point to an index.html file that records the paths of all files.
+    # This allows `index.html` to be displayed directly in HTML format.
+    #
+    # Note: While AWS S3 supports this behavior of rendering HTML files directly,
+    # not all S3-compatible storage services do. For example, Alibaba Cloud OSS
+    # has security restrictions on its default domain that may require files to be
+    # downloaded rather than rendered directly, thus preventing HTML files from being displayed.
+    attr_reader :url
+
+    def initialize(project)
+      @tag = project.engine_tag
+      @timestamp = project.timestamp
+      @logger = project.logger
+      @repo = project.config['repository']
+      @config = Json.read_json(CONFIG_JSON, @logger)
+
+      @access_key_id = ENV.fetch('AUTOHCK_S3COMPATIBLE_ACCESS_KEY_ID')
+      @secret_access_key = ENV.fetch('AUTOHCK_S3COMPATIBLE_SECRET_ACCESS_KEY')
+
+      @region = @config['region']
+      @bucket_name = @config['bucket_name']
+      @endpoint = @config['endpoint']
+
+      @s3_resource = nil
+      @bucket = nil
+    end
+
+    def html_url; end
+
+    # As of now, the exception handling only logs the errors without any additional recovery or fallback mechanisms.
+    # Future enhancements may include more robust error handling strategies.
+    def handle_exceptions(where)
+      yield
+    rescue Aws::Errors::ServiceError => e
+      @logger.error("S3Compatible Error in #{where}: #{e.message}")
+      false
+    rescue StandardError => e
+      @logger.error("S3Compatible General error in #{where}: #{e.message}")
+      false
+    end
```

Comment:
I see your point. Indeed, there is currently no recovery mechanism implemented. I've noticed that the S3 API itself has some built-in retry and recovery features, so the need for additional recovery mechanisms seems limited. The primary purpose of this implementation is to alert users to preliminary errors, such as misconfigured buckets or incorrect access keys, which can still provide valuable guidance and aid during troubleshooting.
I have combined the rescue for Aws::Errors::ServiceError and StandardError. Is it correct to implement it this way?
def handle_exceptions(where)
  yield
rescue StandardError => e
  @logger.error("S3Compatible Error in #{where}: #{e.message}")
  false
end

Reply:
Doesn't AutoHCK automatically handle the error? If not, perhaps it is better to do it in code common with the other uploader (Dropbox).

Reply:
@akihikodaki I looked into it, and if exceptions are not caught here and are instead handled directly by AutoHCK, the entire program will exit if an error occurs. I noticed that Dropbox handles different errors in different ways; some exceptions are thrown, while others are just logged. To be honest, I don't fully understand the logic behind this handling and whether it relates to certain characteristics of Dropbox itself.

Currently, my approach is to catch all exceptions thrown by S3 but only log them. Do you think this needs some improvement? How could it be improved to be more appropriate?

@akihikodaki @kostyanf14  Aside from this review, I have already made the necessary changes to the others. Please take a moment to look at them as well.

Reply:
`handle_exceptions` can be removed. The default behavior, which is to exit the program with a non-zero status code, is the correct way to handle the error in my opinion. That allows the user to notice the uploaded results are corrupted and to fix it.

Reply:
@akihikodaki I don't agree. Results uploader is not a main feature. Dropbox sometimes fails and we don't have a URL or some files but CI in general passes. If we missing one file this is not critical but terminate CI just because of a network issue is a bad idea.

Reply:
Shouldn't the CI be retried when the uploader fails? It may not be that troublesome to retry the CI if uploader failures are rare.

Reply:
Can try if this is possible. In Dropbox we try several times to reconnect or perform operation

Reply:
For now, I suggest removing `handle_exceptions` for S3 but keeping `handle_exceptions` for Dropbox.
S3 and similar services have great availability and can process so many requests. An exception is likely to be an indication of programming error or misconfiguration.
I'm not that confident about Dropbox though so let's keep `handle_exceptions` for Dropbox; we can revisit it whenever necessary.

Reply:
@akihikodaki Sorry, I disagree.

AutoHCK is no longer just a developer tool. It runs in production deployment. Killing a CI run because of the uploaded exception is unacceptable.
With DropBox, it was a case before we added an exception handler in any case, and it was extremely annoying to find from time to time that the test failed just because of some throttling or unavailability of the service. 
Log the error if it happens. We also have integration with Sentury - trigger event in Sentury in case of S3 exception.

Reply:
I still think an upload failure requires the user's attention. S3 and similar services have high availability and can simultaneously serve multiple users while a Dropbox bucket is presumably designed for a few users. Both of throttling and unavailability should be rare conditions.

Reply:
Even if they are rare, when it happens in production, especially when CI is part of the pipeline, it is very destructive not to handle errors that are coming from auxiliary features. 

Better not crash CI and mark it pass or fail based on the test results, and let the user to handle the case when logs are unavailable because of S3 failure (checking the logs, and other possible actions, checking workspace, and etc)

Reply:
I'm still not convinced it's prohibitive even if the events are rare. I haven't seen any occasion where a GitHub Actions/GitLab CI task fails with artifact uploads except that the backend is misconfigured or the network is failing, which requires attention.

Reply:
We had several failures:
1. DropBox API issues
2.  Github API issues (failure to update the status of the CI due to GitHub API temporary unavailability)

Reply:
I’d like to express my concern as well. Since we actually support S3-compatible storage, not just limited to AWS and Alibaba Cloud, I'm worried that not all S3-compatible storage solutions are equally stable.

Reply:
Relaxed rate limiting and high availability are not just of S3 but common for similar services. For example, if such services have strict rate limits, they cannot be used to build Web services. They are intended to be used as building blocks of complex services and should satisfy requirements for such use cases.

Reply:
Again, I must insist on the 100% resilience of the system on our side. 

Any web system has safeguards against 3rd party API failures.

Reply:
A Web service that cannot store user data is non-functional so the availability of the storage service is critical for resilience. We can say the same for a CI system; a CI system that cannot report the results and store artifacts provides no value and requires intervention to fix the issue.

Reply:
Yes, but the web service that cannot momentarily store your picture will not crash the front end.  And it will not crash the micro service that uses this storage. It will log the error and report the anomaly. 

Crashing hck-ci because we couldn't upload test results (and again, this is not theory) is worse than not crashing HCK-CI and reporting the result as pass or fail. The user that will want to investigate will check what happened. In case of crashing.

Moreover, hck-ci crashes can happen when we upload partial results. Let's say CI ran the first HCK test and wants to upload the test results and update the log file that we have (the status of what tests passed, failed, or pending). It fails. But it might succeed next time. Why would we crash hck-ci in such a case? Even worse - it can be CI run executing for hours, and again, only one test result upload fails. Why would we kill the execution and keep the user with partial results?

Reply:
> Yes, but the web service that cannot momentarily store your picture will not crash the front end. And it will not crash the micro service that uses this storage. It will log the error and report the anomaly.
> 
> Crashing hck-ci because we couldn't upload test results (and again, this is not theory) is worse than not crashing HCK-CI and reporting the result as pass or fail. The user that will want to investigate will check what happened. In case of crashing.

Such a Web service will still fail for that particular request. This is the same for AutoHCK; one CI run will fail but the other may continue running.

> 
> Moreover, hck-ci crashes can happen when we upload partial results. Let's say CI ran the first HCK test and wants to upload the test results and update the log file that we have (the status of what tests passed, failed, or pending). It fails. But it might succeed next time. Why would we crash hck-ci in such a case? Even worse - it can be CI run executing for hours, and again, only one test result upload fails. Why would we kill the execution and keep the user with partial results?

I'm thinking of this feature to report the results.

> Moreover, hck-ci crashes can happen when we upload partial results. Let's say CI ran the first HCK test and wants to upload the test results and update the log file that we have (the status of what tests passed, failed, or pending). It fails. But it might succeed next time. Why would we crash hck-ci in such a case? Even worse - it can be CI run executing for hours, and again, only one test result upload fails. Why would we kill the execution and keep the user with partial results?

That is the concern of the caller so it's better to be handled there, not at individual uploader implementations as suggested in: https://github.com/HCK-CI/AutoHCK/pull/476#discussion_r1797697163

Reply:
> That is the concern of the caller so it's better to be handled there, not at individual uploader implementations as suggested in: https://github.com/HCK-CI/AutoHCK/pull/476#discussion_r1797697163

So until it is done (let's open an issue), I suggest having exception handling in the individual uploader.

Reply:
For now, I suggest rescuing `Aws::Errors::ServiceError` and `Seahorse::Client::NetworkingError`. Rescuing `StandardError` will suppress even programming errors like `NoMethodError`. Rescuing these two should suppress errors due to networking and ones reported by the storage service.
We can amend it to raise a special error class instead of logging when we need to handle errors in callers.

Reply:
Okay, I've made the change